evm/src/cpu/jumps.rs

```diff
@@ -0,0 +1,376 @@
+use once_cell::sync::Lazy;
+use plonky2::field::extension::Extendable;
+use plonky2::field::packed::PackedField;
+use plonky2::field::types::Field;
+use plonky2::hash::hash_types::RichField;
+use plonky2::iop::ext_target::ExtensionTarget;
+
+use crate::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+use crate::cpu::columns::CpuColumnsView;
+use crate::cpu::kernel::aggregator::KERNEL;
+
+static INVALID_DST_HANDLER_ADDR: Lazy<usize> =
+    Lazy::new(|| KERNEL.global_labels["fault_exception"]);
+
+pub fn eval_packed_exit_kernel<P: PackedField>(
+    lv: &CpuColumnsView<P>,
+    nv: &CpuColumnsView<P>,
+    yield_constr: &mut ConstraintConsumer<P>,
+) {
+    let jumps_lv = lv.general.jumps();
+
+    // If we are executing `EXIT_KERNEL` then we simply restore the program counter and kernel mode
+    // flag. The top 6 (32-bit) limbs are ignored (this is not part of the spec, but we trust the
+    // kernel to set them to zero).
+    yield_constr.constraint_transition(
+        lv.is_cpu_cycle * lv.is_exit_kernel * (jumps_lv.input0[0] - nv.program_counter),
+    );
+    yield_constr.constraint_transition(
+        lv.is_cpu_cycle * lv.is_exit_kernel * (jumps_lv.input0[1] - nv.is_kernel_mode),
+    );
+}
+
+pub fn eval_ext_circuit_exit_kernel<F: RichField + Extendable<D>, const D: usize>(
+    builder: &mut plonky2::plonk::circuit_builder::CircuitBuilder<F, D>,
+    lv: &CpuColumnsView<ExtensionTarget<D>>,
+    nv: &CpuColumnsView<ExtensionTarget<D>>,
+    yield_constr: &mut RecursiveConstraintConsumer<F, D>,
+) {
+    let jumps_lv = lv.general.jumps();
+    let filter = builder.mul_extension(lv.is_cpu_cycle, lv.is_exit_kernel);
+
+    // If we are executing `EXIT_KERNEL` then we simply restore the program counter and kernel mode
+    // flag. The top 6 (32-bit) limbs are ignored (this is not part of the spec, but we trust the
+    // kernel to set them to zero).
+
+    let pc_constr = builder.sub_extension(jumps_lv.input0[0], nv.program_counter);
+    let pc_constr = builder.mul_extension(filter, pc_constr);
+    yield_constr.constraint_transition(builder, pc_constr);
+
+    let kernel_constr = builder.sub_extension(jumps_lv.input0[1], nv.is_kernel_mode);
+    let kernel_constr = builder.mul_extension(filter, kernel_constr);
+    yield_constr.constraint_transition(builder, kernel_constr);
+}
+
+pub fn eval_packed_jump_jumpi<P: PackedField>(
+    lv: &CpuColumnsView<P>,
+    nv: &CpuColumnsView<P>,
+    yield_constr: &mut ConstraintConsumer<P>,
+) {
+    let jumps_lv = lv.general.jumps();
+    let filter = lv.is_jump + lv.is_jumpi; // `JUMP` or `JUMPI`
+
+    // If `JUMP`, re-use the `JUMPI` logic, but setting the second input (the predicate) to be 1.
+    yield_constr.constraint(lv.is_jump * (jumps_lv.input1[0] - P::ONES));
+    for &limb in &jumps_lv.input1[1..] {
+        // Set all limbs (other than the least-signigficant limb) to 0.
+        // NB: Technically, they don't have to be 0, as long as the sum
```

Comment:
Don't they need to be zero (as you have it) to avoid the fault condition? So a dishonest prover can't cause a JUMP to fault.

Reply:
The fault depends on the address (`input0`). This is setting the condition to 1 (`input1 = 1`). In other words, it implements `JUMP(addr)` as `JUMPI(addr=addr, condition=1)`